### PR TITLE
Foundations for video analytics

### PR DESCRIPTION
**Pull request by PawelPeczek-Roboflow**
# Description

In this PR we are adding bunch of blocks that are important for video analytics:
* data aggregator - to aggregate results of workflow processing over time
* CSV formatter - to format CSV file out of processing results
* File sink - to dump data into file
* Email notification - to send emails in reaction to things happening in workflow
* Webhook sink - to send requests with workflow data 
* Rate limiter - control flow block by @yeldarby to limit the execution pace
* tracker visualizer
* path visualizer (to be able to visualise the input for path deviation block)

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

* E2E
* CI still 🟢 
* new tests in CI

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: requirements/_requirements.txt [hunk 1]
```diff
@@ -31,3 +31,4 @@ shapely>=2.0.0,<2.1.0
 tldextract~=5.1.2
 packaging~=24.0
 anthropic~=0.34.2
+pandas>=2.0.0,<2.1.0
```
Comment (PawelPeczek-Roboflow) [inline requirements/_requirements.txt:34]: @grzegorz-roboflow - added this pin, as from `2.2.0` onward py 3.8 is dropped

**Comment by PawelPeczek-Roboflow** [comment]
tests on Jetson Jetpack 5.1
image

File: inference/core/env.py [hunk 2]
```diff
@@ -442,3 +442,6 @@
 USE_FILE_CACHE_FOR_WORKFLOWS_DEFINITIONS = str2bool(
     os.getenv("USE_FILE_CACHE_FOR_WORKFLOWS_DEFINITIONS", "True")
 )
+ALLOW_WORKFLOW_BLOCKS_ACCESSING_LOCAL_STORAGE = str2bool(
+    os.getenv("ALLOW_WORKFLOW_BLOCKS_ACCESSING_LOCAL_STORAGE", "True")
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/env.py:446]: Was thinking about that as this seemed to me as safer option - the only problem is that running in Python package you would need to manually enable the block - which made me to explicitly disable in lambda builds

File: inference/core/workflows/core_steps/common/query_language/entities/operations.py [hunk 3]
```diff
@@ -378,6 +395,40 @@ class ExtractImageProperty(OperationDefinition):
...
+        },
+    )
+    type: Literal["ConvertImageToJPEG"]
+    compression_level: int = Field(default=95, le=100, ge=1)
+
+
+class ConvertDictionaryToJSON(OperationDefinition):
+    model_config = ConfigDict(
+        json_schema_extra={
+            "description": "Converts dictionary to JSON",
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/common/query_language/entities/operations.py:413]: 👍

File: inference/core/workflows/core_steps/formatters/csv/v1.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    "predictions": [
+        {"type": "DetectionsPropertyExtract", "property_name": "class_name"}
+    ],
+}
+```
+
+Will generate CSV content:
+```csv
+timestamp,predictions,reference
+"2024-10-16T11:15:15.336322","['a', 'b', 'c']","['a', 'b']"
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/formatters/csv/v1.py:70]: 👍

File: inference/core/workflows/core_steps/formatters/csv/v1.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    columns_operations: Dict[str, List[AllOperationsType]],
+) -> List[dict]:
+    result = []
+    for columns_data in unfold_parameters(batch_columns_data=batch_columns_data):
+        for variable_name, operations in columns_operations.items():
+            operations_chain = build_operations_chain(operations=operations)
+            columns_data[variable_name] = operations_chain(
+                columns_data[variable_name], global_parameters={}
+            )
+        columns_data["timestamp"] = datetime.now().isoformat()
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/formatters/csv/v1.py:231]: 👍

File: inference/core/workflows/core_steps/sinks/email_notification/v1.py [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+Message parameters are delivered by Workflows Execution Engine by setting proper data selectors in
+`message_parameters` field, for example:
+
+```
+message_parameters = {
+    "predicted_classes": "$steps.model.predictions"
+}
+```
+
+Selecting data is not the only option - data may be processed in the block. In the example we wish to
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/sinks/email_notification/v1.py:74]: 👍

File: inference/core/workflows/core_steps/sinks/email_notification/v1.py [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+        ],
+        default_factory=dict,
+    )
+    attachments: Dict[str, StepOutputSelector(kind=[STRING_KIND])] = Field(
+        description="Attachments",
+        default_factory=dict,
+        examples=[{"report.cvs": "$steps.csv_formatter.csv_content"}],
+    )
+    smtp_server: Union[str, WorkflowParameterSelector(kind=[STRING_KIND])] = Field(
+        description="Custom SMTP server to use",
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/sinks/email_notification/v1.py:244]: 👍

File: inference/core/workflows/core_steps/sinks/email_notification/v1.py [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+        examples=["$inputs.smtp_server", "smtp.google.com"],
+    )
+    sender_email_password: Union[str, WorkflowParameterSelector(kind=[STRING_KIND])] = (
+        Field(
+            description="Sender e-mail password to use SMTP server",
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/sinks/email_notification/v1.py:249]: 👍

File: inference/core/workflows/core_steps/sinks/email_notification/v1.py [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+            private=True,
+            examples=["$inputs.email_password"],
+        )
+    )
+    smtp_port: int = Field(
+        default=465,
+        description="Port of custom SMTP server to use",
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/sinks/email_notification/v1.py:256]: 👍

File: inference/core/workflows/core_steps/sinks/local_file/v1.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+        if self._active_file_descriptor is not None:
+            self._active_file_descriptor.close()
+
+
+def generate_new_file_path(
+    target_directory: str, file_name_prefix: str, file_type: str
+) -> str:
+    timestamp = datetime.now().strftime("%Y_%m_%d_%H_%M_%S_%f")
+    file_name = f"{file_name_prefix}_{timestamp}.{file_type}"
+    return os.path.abspath(os.path.join(target_directory, file_name))
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/sinks/local_file/v1.py:282]: let's discuss that

File: inference/core/workflows/core_steps/transformations/byte_tracker/v3.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        not_seen_instances_detections = tracked_detections[not_seen_instances_mask]
+        already_seen_instances_detections = tracked_detections[seen_instances_mask]
+        return {
+            OUTPUT_KEY: tracked_detections,
+            "new_instances": not_seen_instances_detections,
+            "already_seen_instances": already_seen_instances_detections,
+        }
+
+
+class InstanceCache:
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/transformations/byte_tracker/v3.py:194]: for now this abstraction is fine, as we would use this only to cache detections of specific tracker

File: inference/core/workflows/core_steps/visualizations/trace/v1.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                [
+                    color_palette,
+                    palette_size,
+                    color_axis,
+                    position,
+                    trace_length,
+                    thickness,
+                ],
+            )
+        )
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/visualizations/trace/v1.py:145]: let's discuss that

**Review by grzegorz-roboflow — APPROVED** [review]

**Pull request merged**